### DISCUSSION TITLE
Run Containerlab in a GitLab-CI Docker-in-Docker environment

### DISCUSSION THREAD

**Discussion by neilschark** [srl-labs/containerlab#1688] [Q&A]
Hi,

I got containerlab running in a docker-in-docker setup without problems with a VSCode dev container setup as it is mentioned in the documentation of containerlab.

The next step for us is to have containerlab running in GitLab-CI. We currently use a shell runner for it, but this setup is not very good for obvious reasons. So we want to migrate it so that each test uses its own containerized instance of containerlab and a topology.

I configured a docker-in-docker runner and  based my image on the default docker-in-docker alpine image.
I then added all dependencies listed in containerlabs own dockerfile: https://github.com/srl-labs/containerlab/blob/29bf4912f127d5ccd28e71a59d6938046ac0755b/goreleaser.dockerfile#L6C1-L15C17

With this image, docker runs fine:
```sh
$ docker run hello-world
Unable to find image 'hello-world:latest' locally
latest: Pulling from library/hello-world
719385e32844: Pulling fs layer
719385e32844: Verifying Checksum
719385e32844: Download complete
719385e32844: Pull complete
Digest: sha256:88ec0acaa3ec199d3b7eaf73588f4518c25f9d34f58ce9a0df68429c5af48e8d
Status: Downloaded newer image for hello-world:latest
Hello from Docker!
[...]
```

Docker can also create networks:
```sh
$ docker network create -d bridge my-bridge-network
3734c8d908be1011ea1addc0d00df57747d899499c38b7b3e60f9c43003fbbe1
```

But unfortunately containerlab isn't able to create the docker network:
```sh
$ containerlab deploy --log-level trace -t dev_env_data/clab/basic_two_aristas.yaml
[...]
time="2023-11-01T09:47:05Z" level=debug msg="Error opening host DNS config run/systemd/resolve/resolv.conf: open run/systemd/resolve/resolv.conf: no such file or directory"
time="2023-11-01T09:47:05Z" level=debug msg="lab Conf: &{Name:gosdn_two_aristas Prefix:0xc00057c1a0 Mgmt:0xc000220090 Settings:<nil> Topology:0xc0006a4180 Debug:false}"
time="2023-11-01T09:47:05Z" level=debug msg="Env: CLAB_VERSION_CHECK="
time="2023-11-01T09:47:05Z" level=debug msg="Checking if docker network \"gosdn-net\" exists"
time="2023-11-01T09:47:05Z" level=debug msg="Network \"gosdn-net\" does not exist"
time="2023-11-01T09:47:05Z" level=info msg="Creating docker network: Name=\"gosdn-net\", IPv4Subnet=\"172.100.0.0/16\", IPv6Subnet=\"2001:db8::/64\", MTU='ל'"
time="2023-11-01T09:47:06Z" level=warning msg="failed gleaning v4 and/or v6 addresses from bridge via netlink, falling back to docker network inspect data"
Error: failed to lookup link "br-c785b71806d6": Link not found
```

I then tried using the containerlab docker image directly as my image for this CI job. But this doesn't work, as GitLab-CI doesn't mount the docker socket into the container, but rather exposes the API to the container as a network service: https://docs.gitlab.com/ee/ci/docker/using_docker_build.html#use-docker-in-docker

Which leads to the following obvious error:
```sh
$ containerlab deploy --log-level trace -t dev_env_data/clab/basic_two_aristas.yaml
[...]
Error: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
```

How is it possible to run containerlab in a GitLab-CI docker-in-docker environment?

**Comment by hellt** [comment]
Hi @neilschark 
can you try using the d-in-d image we use in clabernetes to run docker-in-docker with containerlab? The dockerfile is here https://github.com/srl-labs/clabernetes/blob/main/build/launcher.Dockerfile

/cc @carlmontanari

**Reply by carlmontanari** [reply]
hmm feels similar to some stuff we saw on discord -- iirc the first error happens when we cant find the gateway address of the network, but I never got around to digging more. 

I dont think the image in c9s would do anything special to help here, but you could give it a whack for sure.

**Reply by neilschark** [reply]
Thanks for your replies, I will look into it.

**Reply by neilschark** [reply]
I used the Dockerfile @hellt provided, only added some debian packages we would need and the problem is the same as using the containerlab image, it can't find the docker daemon, as the socket is not mounted, but the API is provided via a service. So as @carlmontanari mentioned, it unfortunately didn't help.

**Reply by hellt** [reply]
@neilschark socket is not required in that case. Docker-in-docker does not leverage the socket bind mounting, since the docker daemon is expected to run in the container, and not on the host.

We noticed that sometimes the docker doesn't start in d-in-d mode when iptables-legacy package is missing.

```
[*]─[test-clab]─[/etc/init.d]
└──> update-alternatives --set iptables /usr/sbin/iptables-legacy
update-alternatives: using /usr/sbin/iptables-legacy to provide /usr/sbin/iptables (iptables) in manual mode

[*]─[test-clab]─[/etc/init.d]
└──> service docker start
Starting Docker: docker.

[*]─[test-clab]─[/etc/init.d]
└──> docker ps
CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES
```

**Reply by neilschark** [reply]
@hellt I am a little confused by your reply. What case do you exactly mean? Den clabernetes image, the containerlab image or my self-build docker-in-docker image with containerlab installed?

Because in my own d-i-d image, docker works fine, just containerlab can't create the network links. In the other two images, the docker socket isn't found, as containerlab looks for `unix:///var/run/docker.sock`, but docker is exposed via a service accessible by the container under the hostname `docker` as described in the GitLab-CI documentation:
```
  # When you use the dind service, you must instruct Docker to talk with
  # the daemon started inside of the service. The daemon is available
  # with a network connection instead of the default
  # /var/run/docker.sock socket.

  # The 'docker' hostname is the alias of the service container as described at
  # https://docs.gitlab.com/ee/ci/services/#accessing-the-services.
```

**Reply by hellt** [reply]
what I meant to say is that there are two modes often abbreviated as docker-in-docker

1. fully d-in-d experience where both docker daemon and docker client are installed inside a container. That way no docker socket is required to be mounted inside the container. This is the "original" docker-in-docker deployment mode.
2. somewhat docker-in-docker where docker client runs inside the container, but daemon is running outside of it. In that mode socket needs to be propagated to the container namespace for a client to function. Note, that in this case containers are actually created not in the d-in-d container, but on the host, because that's where docker daemon is running.

We will have to check what Gitlab CI requires for d-in-d experience.

**Reply by neilschark** [reply]
Thanks four your reply. While I am using Docker for around 6 years now, I am relatively new to docker in docker, so I was unaware of the differences and have only used the second mode you described.

As I understand the docker in docker functionality of GitLab-CI, it also uses the second mode. The docker API is made available as a 'service', which is a network link to another container. https://docs.gitlab.com/ee/ci/services/#how-services-are-linked-to-the-job

The only things mounted to the container running your jobs in the default configuration are `/cache` and `/certs/clients` for TLS.
To access docker, you access its API via the service.

**Reply by neilschark** [reply]
I made some progress:

One can manually set the URL where docker can be found with the env var called `DOCKER_HOST`. 
E.g.: `DOCKER_HOST: tcp://docker:2375`.

I tried it with and without TLS and the default clab image. If I install the docker-cli package, `docker info` and `docker run hello-world` fail, if the env var is NOT set, as it tries to use the default socket location. If I then set the env var, docker works. I tested it with and without TLS.

Containerlab apparently tries to use the env war as well, because event though it fails with the same error message, containerlab seems to try to access the given path to the API.
```
Error: Cannot connect to the Docker daemon at tcp://docker:2375/. Is the docker daemon running?
```

I am a little confused by this. To summarize:

If I use the official docker in docker image, containerlab fails without a good error message. If I use the official containerlab image and change the configuration in a way, that docker works, containerlab can't access the docker daemon, even though it lists the correct URL for it.

**Reply by hellt** [reply]
yes, in contianerlab's docker client code we respect the env vars known to docker.
Not sure why it refuses to connect to the daemon over TCP
I bet if you use tcpdump you will requests coming from containerlab towards docker:2375

**Reply by neilschark** [reply]
I will look into running tcpdump on it next week and report back.

**Reply by neilschark** [reply]
I tried using tcpdump to see if containerlab tries to connect to the docker daemon, but apparently because of docker iptables magic, you apparently can't simply use tcpdump in the same container. You have to connect a separate container into the same docker network. This is a relatively complicated task simply to see if containerlab tries to connect to the docker daemon, as the main problem would likely remain. As is shown by my self-build docker image, even if containerlab can connect, it still can't create the network.

Do any of you still have an idea how I can proceed and get containerlab running in GitLab-CI with docker-in-docker?

**Reply by hellt** [reply]
I haven't tried dind in gitlab ci.
Do you use self-hosted runners with gilab ci?

**Reply by neilschark** [reply]
Yes, I use a self-hosted runner. It is a root server where I have full root access. The server is running all our pipelines for the project.

But I have access to servers on demand, so I could also create a different GitLab-Runner just for testing, if some configurations should be tested which could potentially break the runner, so that my coworkers are not disturbed in their work.
